src/workers/python/papyros/papyros.py

```diff
@@ -0,0 +1,164 @@
+import os
+import sys
+import json
+import python_runner
+import friendly_traceback
+
+from friendly_traceback.core import FriendlyTraceback
+from collections.abc import Awaitable
+from pyodide_worker_runner import install_imports
+
+
+from .util import to_py
+from .autocomplete import autocomplete
+
+SYS_RECURSION_LIMIT = 500
+
+# Global Papyros instance
+papyros = None
+
+class Papyros(python_runner.PyodideRunner):
+    def __init__(
+        self,
+        *,
+        source_code="",
+        filename="my_program.py",
+        callback=None,
+        limit=SYS_RECURSION_LIMIT
+    ):
+        if callback is None:
+            raise ValueError("Callback must not be None")
+        super().__init__(source_code=source_code, filename=filename)
+        self.limit = limit
+        self.override_globals()
+        self.set_event_callback(callback)
+
+    def set_event_callback(self, event_callback):
+        def runner_callback(event_type, data):
+            def cb(typ, dat, contentType=None, **kwargs):
+                return event_callback(dict(type=typ, data=dat, contentType=contentType or "text/plain", **kwargs))
+
+            if event_type == "output":
+                for part in data["parts"]:
+                    typ = part["type"]
+                    if typ in ["stderr", "traceback", "syntax_error"]:
+                        cb("error", part["text"], contentType=part.get("contentType"))
+                    elif typ in ["input", "input_prompt"]:
+                        # Do not display values entered by user for input
+                        continue
+                    else:
+                        cb("output", part["text"], contentType=part.get("contentType"))
```

Comment:
I think you should embrace the python_runner output part types approach. You're sending stderr and stdout to separate event types and code paths, but they're almost exactly the same thing, just different colours. Meanwhile you're combining stderr and tracebacks which get handled completely differently. Ultimately these are all just different types of output which all get rendered to the same output area, even if it's in different ways.

Reply:
These things are a bit too tied to Python, so they're not easily generalised to other languages (e.g. due to presence of snake_case). With the eventing system, I think it's nice to be able to only listen to error messages instead of getting all output. Finally, I don't think our target group will often write to stderr, so the focus is on the tracebacks as they will be the main reason for error displaying. The current translations seem fine to me.